rs/sns/governance/src/governance.rs

```diff
@@ -5541,93 +5562,122 @@ impl Governance {
                 });
 
-                // Reset cached_upgrade_steps.
-                let cached_upgrade_steps =
-                    CachedUpgradeSteps::empty(running_version, self.env.now());
-
-                self.proto
-                    .cached_upgrade_steps
-                    .replace(CachedUpgradeStepsPb::from(cached_upgrade_steps));
-
                 self.fail_sns_upgrade_to_next_version_proposal(
                     proposal_id,
-                    GovernanceError::new_with_message(ErrorType::PreconditionFailed, error),
+                    GovernanceError::new_with_message(ErrorType::External, message),
                 );
-                return;
             }
-            Ok(cached_upgrade_steps) => cached_upgrade_steps,
-        };
-
-        let expected_changes = current_version.changes_against(&target_version);
 
-        match running_version.version_has_expected_hashes(&expected_changes) {
-            Ok(_) => {
-                log!(
-                    INFO,
-                    "Upgrade marked successful at {} from genesis.  New Version: {:?}",
-                    self.env.now(),
-                    target_version
-                );
-                self.push_to_upgrade_journal(upgrade_journal_entry::UpgradeOutcome {
-                    status: Some(upgrade_journal_entry::upgrade_outcome::Status::Success(
-                        Empty {},
-                    )),
-                    human_readable: None,
-                });
-                self.set_proposal_execution_status(proposal_id, Ok(()));
+            // Do not update the cached upgrade steps since an upgrade did not happen (yet).
+            return;
+        }
 
+        // Update the cached upgrade steps to reflect the fact the 1st version is no longer needed.
+        //
+        // The rest of the code in this function should comply with the following specification:
+        //
+        // Case I:
+        //     old state (before previous await):
+        //         cached_upgrade_steps: [v1, v2, v3, v4]
+        //         v1: running_version -> v2: target_version
+        //
+        //     new state:
+        //         v2: running_version == target_version
+        //         cached_upgrade_steps: [v1]                  // notice that this is a singleton
+        //
+        //     cached_upgrade_steps' == [v2]                   // reset
+        //
+        // Case II:
+        //     old state (before previous await):
+        //         cached_upgrade_steps: [v1, v2, v3, v4]
+        //         v1: running_version -> v2: target_version
+        //
+        //     new state:
+        //         v2: running_version == target_version
+        //         cached_upgrade_steps: [v1, v2', ...]
+        //
+        //     Case II(a): v2' == v2:
+        //         cached_upgrade_steps' == [v2, ...]            // consume
+        //
+        //     Case II(b): v2' != v2:
+        //         cached_upgrade_steps' == [v2]                 // reset
```

Comment:
Did you mean to include this comment? I find it less clear than the code TBH

Reply:
I think we should drop it before merging. We don't have a standard convention for writing stateful specs in the code, unfortunately.